### PR TITLE
optimize convertIstioListenerToWrapper performance

### PR DESCRIPTION
**Pull request by wulianglongrd**
update on 8.25:

Save the exact hosts(Sidecar CR Egress.Hosts) separately in Set (internal implementation is map), so that in the following process can be matched using map search instead of loop.

- When hosts are all exact hosts: performance will be improved. In our production environment, the average time-consuming for[ this indicator](https://github.com/istio/istio/blob/master/pilot/pkg/xds/monitoring.go#L105) has dropped by 50%+ (off-topic: I think the indicator bounds are set too narrowly, there is a [discussion here](https://github.com/istio/istio/pull/45924)). From the benchmarks below, the performance improvement is not obvious, which may be caused by inappropriate benchmark test case data.

- When hosts are all wildcard hosts: performance may drop, but not much (<5%).

The memory will be slightly increased (<10%), but it seems to be acceptable, because too much cpu consumption is the current problem.

**benchmark result(update on 8.25)**:
```shell
$ go test -run="none" -bench="BenchmarkConvertIstioListenerToWrapper" -test.benchmem -count=5 . > old.txt
$ go test -run="none" -bench="BenchmarkConvertIstioListenerToWrapper" -test.benchmem -count=5 . > new.txt
$ benchstat old.txt new.txt
name                                               old time/op    new time/op    delta
ConvertIstioListenerToWrapper/small-exact-12         3.79µs ± 4%    3.86µs ± 7%     ~     (p=0.841 n=5+5)
ConvertIstioListenerToWrapper/small-wildcard-12      2.23µs ± 1%    2.30µs ± 2%   +3.22%  (p=0.008 n=5+5)
ConvertIstioListenerToWrapper/small-match-all-12     6.38µs ±16%    6.41µs ± 2%     ~     (p=0.310 n=5+5)
ConvertIstioListenerToWrapper/middle-exact-12        52.3µs ±23%    37.6µs ± 0%  -28.13%  (p=0.016 n=5+4)
ConvertIstioListenerToWrapper/middle-wildcard-12     40.7µs ± 7%    36.7µs ± 1%   -9.78%  (p=0.008 n=5+5)
ConvertIstioListenerToWrapper/middle-match-all-12    55.7µs ± 5%    54.3µs ± 3%     ~     (p=0.421 n=5+5)
ConvertIstioListenerToWrapper/big-exact-12            183µs ± 1%     147µs ±10%  -20.09%  (p=0.016 n=4+5)
ConvertIstioListenerToWrapper/big-wildcard-12         168µs ±30%     140µs ± 2%  -16.62%  (p=0.008 n=5+5)
ConvertIstioListenerToWrapper/big-match-all-12        216µs ±49%     177µs ± 3%     ~     (p=0.310 n=5+5)

name                                               old alloc/op   new alloc/op   delta
ConvertIstioListenerToWrapper/small-exact-12         2.45kB ± 0%    2.64kB ± 0%   +7.84%  (p=0.008 n=5+5)
ConvertIstioListenerToWrapper/small-wildcard-12        592B ± 0%      640B ± 0%   +8.11%  (p=0.008 n=5+5)
ConvertIstioListenerToWrapper/small-match-all-12     9.69kB ± 0%    9.74kB ± 0%   +0.50%  (p=0.008 n=5+5)
ConvertIstioListenerToWrapper/middle-exact-12        13.4kB ± 0%    13.9kB ± 0%   +3.60%  (p=0.016 n=5+4)
ConvertIstioListenerToWrapper/middle-wildcard-12     4.18kB ± 0%    4.22kB ± 0%   +1.15%  (p=0.008 n=5+5)
ConvertIstioListenerToWrapper/middle-match-all-12    88.1kB ± 0%    88.2kB ± 0%   +0.05%  (p=0.008 n=5+5)
ConvertIstioListenerToWrapper/big-exact-12           22.0kB ± 0%    23.1kB ± 0%   +4.98%  (p=0.008 n=5+5)
ConvertIstioListenerToWrapper/big-wildcard-12        10.9kB ± 0%    11.0kB ± 0%   +0.44%  (p=0.008 n=5+5)
ConvertIstioListenerToWrapper/big-match-all-12        350kB ± 0%     350kB ± 0%   +0.01%  (p=0.008 n=5+5)

name                                               old allocs/op  new allocs/op  delta
ConvertIstioListenerToWrapper/small-exact-12           27.0 ± 0%      29.0 ± 0%   +7.41%  (p=0.008 n=5+5)
ConvertIstioListenerToWrapper/small-wildcard-12        18.0 ± 0%      19.0 ± 0%   +5.56%  (p=0.008 n=5+5)
ConvertIstioListenerToWrapper/small-match-all-12       31.0 ± 0%      32.0 ± 0%   +3.23%  (p=0.008 n=5+5)
ConvertIstioListenerToWrapper/middle-exact-12           134 ± 0%       137 ± 0%   +2.24%  (p=0.008 n=5+5)
ConvertIstioListenerToWrapper/middle-wildcard-12        117 ± 0%       118 ± 0%   +0.85%  (p=0.008 n=5+5)
ConvertIstioListenerToWrapper/middle-match-all-12       145 ± 0%       146 ± 0%   +0.69%  (p=0.008 n=5+5)
ConvertIstioListenerToWrapper/big-exact-12              341 ± 0%       345 ± 0%   +1.17%  (p=0.008 n=5+5)
ConvertIstioListenerToWrapper/big-wildcard-12           322 ± 0%       323 ± 0%   +0.31%  (p=0.008 n=5+5)
ConvertIstioListenerToWrapper/big-match-all-12          363 ± 0%       363 ± 0%     ~     (p=0.159 n=5+5)

### CODE REVIEW

File: pilot/pkg/model/sidecar_test.go [hunk 1]
```diff
@@ -2517,7 +2538,7 @@ func benchmarkConvertIstioListenerToWrapper(b *testing.B, vsNum int, hostNum int
 		})
 	}
 	ps := NewPushContext()
-	ps.virtualServiceIndex.publicByGateway["default"] = cfgs
```
Comment (wulianglongrd) [inline pilot/pkg/model/sidecar_test.go:2520]: it is a bug, see here

https://github.com/istio/istio/blob/897ae22b5019ead245ae26553dd4dd5ec45c50f6/pilot/pkg/model/virtualservice.go#L82

File: pilot/pkg/model/sidecar_test.go [hunk 2]
```diff
@@ -2095,73 +2096,93 @@ func TestIstioEgressListenerWrapper(t *testing.T) {
...
 		namespace     string
 	}{
 		{
-			name:          "*/* imports only those in a",
-			listenerHosts: map[string][]host.Name{wildcardNamespace: {wildcardService}},
-			services:      allServices,
-			expected:      []*Service{serviceA8000, serviceA9000, serviceAalt},
-			namespace:     "a",
+			name: "*/* imports only those in a",
+			listenerHosts: map[string]hostClassification{
```
Comment (wulianglongrd) [inline pilot/pkg/model/sidecar_test.go:2106]: Only `listenerHosts` has been modified here, and other fields have not been modified, it is format diff.

**Review by howardjohn — COMMENTED** [review]
Looks fine. Surprised this is important - how many entries do you have here in real world?

**Comment by hzxuzhonghu** [comment]
Is the optimization mostly based on host match? If so i would ask how many hosts do we usually set on sidecar? 

I doubt the complexity introduced weigh over the perf improve

**Comment by wulianglongrd** [comment]
> Is the optimization mostly based on host match? If so i would ask how many hosts do we usually set on sidecar?

Yes, a Sidecar CR's egress usually contains 100~400 hosts. We are currently not using wildcard host.

> I doubt the complexity introduced weigh over the perf improve

I'm also a little concern. This PR currently has a bug (vs.host is a wildcard scenario not considered). I'll take another look.

**Comment by hzxuzhonghu** [comment]
>Yes, a Sidecar CR's egress usually contains 100~400 hosts. We are currently not using wildcard host.

Confusing to me, what is your application do while needing to access more than 100 service

**Comment by wulianglongrd** [comment]
forgot there was a discussion here, so sorry.

> how many entries do you have here in real world?

hello @howardjohn , about 50. It is not only related to the number of SEs, but also related to the number of `SE.hosts`.


> Confusing to me, what is your application do while needing to access more than 100 service

hello @hzxuzhonghu , there are not so many real services, 5~20 are more common. This is mainly caused by the large size of `SE.hosts`. Each service currently using mesh has about 15 clusters (some clusters are deployed independently for specific usage scenarios), each with a unique host. Currently, our self-developed tools for tracking service dependencies in real time are not yet mature(use it to update `Sidecar.egress.hosts` in real time). Therefore, if a consumer depends on a provider, we will write all clusters (hosts) of the provider to `Sidecar.egress.hosts`. There are other reasons, like we have multiple namespaces for scope control, etc.

File: pilot/pkg/model/sidecar.go [hunk 3]
```diff
@@ -56,6 +56,11 @@ var (
 	)
 )
 
+type hostClassification struct {
```
[the remark this answers is not in the document]
Comment (wulianglongrd) [reply pilot/pkg/model/sidecar.go:59]: Noticed this issue, but I don't have a good solution at the moment, any suggestions?

**Comment by sschepens** [comment]
@wulianglongrd any chance of getting this in before `1.19`? we're seeing big time spent on this as well

**Comment by wulianglongrd** [comment]
> @wulianglongrd any chance of getting this in before `1.19`? we're seeing big time spent on this as well

Hello @sschepens , I have pushed the code that our production environment actually uses to here. I don't have a better way to optimize it. If you have a good idea, welcome to discuss it here, thank you.

**Comment by sschepens** [comment]
> Hello @sschepens , I have pushed the code that our production environment actually uses to here. I don't have a better way to optimize it. If you have a good idea, welcome to discuss it here, thank you.

@wulianglongrd does it need further optimization? currently we're seeing aproximately 15%  of cpu usage go to `istio.io/istio/pilot/pkg/model.matchingService` and 30% to `istio.io/istio/pilot/pkg/model.SelectVirtualServices` which is exactly what you're optimizing. We use non wildcard hosts so I' assuming the performance increase should be significant in our case since it's the best case scenario.

**Comment by wulianglongrd** [comment]
This may be a compromise between efficiency and complexity.

I think @howardjohn's [comment ](https://github.com/istio/istio/pull/46043#discussion_r1265601483)makes sense:
> We can probably get a lot more efficient actually, using a trie like structure. It might be worthwhile to abstract this to more places.

I also agree with @hzxuzhonghu 's comment that it introduces a lot of complexity and reduces readability.

**Comment by wulianglongrd** [comment]
Not sure if further optimization is needed. Might be possible in the next PR :)

**Comment by sschepens** [comment]
Trie structures are typically costlier to build, we would have to check if the cost of building it would be outweighed by the benefits.

**Review by howardjohn — APPROVED** [review]

**Pull request merged**